File: simo-player/tools/sumo_geom.py
#!/usr/bin/env python3
"""Shared geometry extraction from SUMO .net.xml.

Extracted from sim/build_player.py:geom() so both the case-study build
and the contributor packer use identical logic.
"""
import os
import re
import sys


def find_sumo():
    """Discover SUMO binary: $SUMO_HOME -> macOS framework -> PATH."""
    for c in (
        os.environ.get("SUMO_HOME", "") + "/bin/sumo",
        "/Library/Frameworks/EclipseSUMO.framework/Versions/Current/EclipseSUMO/share/sumo/bin/sumo",
        "sumo",
    ):
        if c and os.path.isfile(c) and os.access(c, os.X_OK):
            return c
    return None


def geo_lock(net_path):
    """<location> -> (latlngMap, utm), mirroring app.js parseNetXml.

    latlngMap: {"conv": convBoundary metres, "orig": origBoundary reordered to
    [minLat, minLng, maxLat, maxLng]}. utm: {"offX","offY","zone","south"}.
    Each is None when the net lacks usable provenance (hand nets write
    -1e10 sentinels)."""
    with open(net_path, encoding="utf-8") as f:
        s = f.read()
    m = re.search(r"<location\s[^>]*>", s)
    if not m:
        return None, None
    tag = m.group(0)

    def attr(name):
        mm = re.search(rf'{name}="([^"]+)"', tag)
        return mm.group(1) if mm else None

    def bbox4(v):
        if not v:
            return None
        f = [float(x) for x in v.split(",")]
        return f if len(f) == 4 else None

    def vec2(v):
        if not v:
            return None
        f = [float(x) for x in v.split(",")]
        return f if len(f) == 2 else None

    convB = bbox4(attr("convBoundary"))
    origB = bbox4(attr("origBoundary"))
    netOff = vec2(attr("netOffset"))
    proj = attr("projParameter") or ""

    latlng_map = None
    if (convB and origB and all(abs(v) <= 1000 for v in origB)
            and convB[2] > convB[0] and convB[3] > convB[1]
            and origB[2] > origB[0] and origB[3] > origB[1]):
        latlng_map = {"conv": convB,
                      "orig": [origB[1], origB[0], origB[3], origB[2]]}

    utm = None
    zm = re.search(r"\+zone=(\d+)", proj)
    if netOff and "+proj=utm" in proj and zm and 1 <= int(zm.group(1)) <= 60:
        utm = {"offX": netOff[0], "offY": netOff[1],
               "zone": int(zm.group(1)), "south": "+south" in proj}
    return latlng_map, utm


def geom(net_path):
    """Extract player geometry from a SUMO .net.xml.

    Returns dict with keys: lanes, arms, phases, stops, links.
    Lane coordinates converted to decimetres (dm) to match stream format.
    """
    with open(net_path, encoding="utf-8") as f:
        s = f.read()

    lanes = []
    for m in re.finditer(r'<lane id="([^:][^"]*)"([^>]*)>?', s):
        attrs = m.group(2)
        sh = re.search(r'shape="([^"]+)"', attrs)
        if not sh:
            continue
        w = re.search(r'width="([\d.]+)"', attrs)
        pts = [[round(float(v) * 10) for v in q.split(",")]
               for q in sh.group(1).split()]
        lanes.append({"p": pts, "w": float(w.group(1)) if w else 3.2})

    ends = {}
    for m in re.finditer(r'<junction id="([^:][^"]*)" type="dead_end" '
                         r'x="([-\d.]+)" y="([-\d.]+)"', s):
        ends[m.group(1)] = [round(float(m.group(2)) * 10),
                            round(float(m.group(3)) * 10)]

    arms = {}
    if ends:
        xs = sorted(ends.values(), key=lambda v: v[0])
        ys = sorted(ends.values(), key=lambda v: v[1])
        arms = {"Panathur": xs[0], "Varthur": xs[-1],
                "Sarjapur": ys[0], "Kundalahalli": ys[-1]}

    phases, stops, links = [], {}, {}

    if "<tlLogic" in s:
        phases = [[float(m.group(1)), m.group(2)] for m in
                  re.finditer(r'<phase duration="([\d.]+)" state="(\w+)"', s)]
        for m in re.finditer(r'<connection from="([^"]+)"[^>]*tl="[^"]*"'
                             r'[^>]*linkIndex="(\d+)"', s):
            frm, idx = m.group(1), int(m.group(2))
            links.setdefault(frm, []).append(idx)
            lm = re.search(rf'<lane id="{re.escape(frm)}_0"[^>]*shape="([^"]+)"', s)
            if lm:
                last = lm.group(1).split()[-1].split(",")
                stops[frm] = [round(float(last[0]) * 10), round(float(last[1]) * 10)]

    return {"lanes": lanes, "arms": arms, "phases": phases,
            "stops": stops, "links": links}


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: sumo_geom.py <net.xml>", file=sys.stderr)
        sys.exit(1)
    import json
    print(json.dumps(geom(sys.argv[1]), separators=(",", ":")))

File: simo-player/tools/sumo_geom.js
#!/usr/bin/env node
/* Shared geometry extraction from SUMO .net.xml.
 *
 * Direct regex port of tools/sumo_geom.py (itself extracted from
 * sim/build_player.py:geom()) so the contributor packer uses identical
 * logic to the retired Python tooling.
 */
import fs from 'node:fs';

function isExecutableFile(p) {
  try {
    return fs.statSync(p).isFile() && fs.accessSync(p, fs.constants.X_OK) === undefined;
  } catch {
    return false;
  }
}

/* Discover SUMO binary: $SUMO_HOME -> macOS framework -> PATH. */
export function findSumo() {
  const candidates = [
    (process.env.SUMO_HOME || '') + '/bin/sumo',
    '/Library/Frameworks/EclipseSUMO.framework/Versions/Current/EclipseSUMO/share/sumo/bin/sumo',
    'sumo',
  ];
  for (const c of candidates) {
    if (c && isExecutableFile(c)) return c;
  }
  return null;
}

/* <location> -> {latlngMap, utm}, mirroring src/lib/netxml.js parseNetXml.
 *
 * latlngMap: {conv: convBoundary metres, orig: origBoundary reordered to
 * [minLat, minLng, maxLat, maxLng]}. utm: {offX, offY, zone, south}.
 * Each is null when the net lacks usable provenance (hand nets write
 * -1e10 sentinels). */
export function geoLock(netPath) {
  const s = fs.readFileSync(netPath, 'utf8');
  const m = /<location\s[^>]*>/.exec(s);
  if (!m) return { latlngMap: null, utm: null };
  const tag = m[0];

  const attr = (name) => {
    const mm = new RegExp(name + '="([^"]+)"').exec(tag);
    return mm ? mm[1] : null;
  };
  const bbox4 = (v) => {
    if (!v) return null;
    const f = v.split(',').map(parseFloat);
    return f.length === 4 ? f : null;
  };
  const vec2 = (v) => {
    if (!v) return null;
    const f = v.split(',').map(parseFloat);
    return f.length === 2 ? f : null;
  };

  const convB = bbox4(attr('convBoundary'));
  const origB = bbox4(attr('origBoundary'));
  const netOff = vec2(attr('netOffset'));
  const proj = attr('projParameter') || '';

  let latlngMap = null;
  if (convB && origB && origB.every((v) => Math.abs(v) <= 1000)
      && convB[2] > convB[0] && convB[3] > convB[1]
      && origB[2] > origB[0] && origB[3] > origB[1]) {
    latlngMap = { conv: convB,
                  orig: [origB[1], origB[0], origB[3], origB[2]] };
  }

  let utm = null;
  const zm = /\+zone=(\d+)/.exec(proj);
  if (netOff && proj.includes('+proj=utm') && zm
      && 1 <= parseInt(zm[1], 10) && parseInt(zm[1], 10) <= 60) {
    utm = { offX: netOff[0], offY: netOff[1],
            zone: parseInt(zm[1], 10), south: proj.includes('+south') };
  }
  return { latlngMap, utm };
}

/* Extract player geometry from a SUMO .net.xml.
 *
 * Returns object with keys: lanes, arms, phases, stops, links.
 * Lane coordinates converted to decimetres (dm) to match stream format. */
export function geom(netPath) {
  const s = fs.readFileSync(netPath, 'utf8');

  const lanes = [];
  for (const m of s.matchAll(/<lane id="([^:][^"]*)"([^>]*)>?/g)) {
    const attrs = m[2];
    const sh = /shape="([^"]+)"/.exec(attrs);
    if (!sh) continue;
    const w = /width="([\d.]+)"/.exec(attrs);
    /* python: sh.group(1).split() — whitespace-run split, no empties */
    const pts = sh[1].trim().split(/\s+/).map((q) => {
      const v = q.split(',');
      return [Math.round(parseFloat(v[0]) * 10), Math.round(parseFloat(v[1]) * 10)];
    });
    lanes.push({ p: pts, w: w ? parseFloat(w[1]) : 3.2 });
  }

  const ends = {};
  const junctionRe = /<junction id="([^:][^"]*)" type="dead_end" x="([-\d.]+)" y="([-\d.]+)"/g;
  for (const m of s.matchAll(junctionRe)) {
    ends[m[1]] = [Math.round(parseFloat(m[2]) * 10),
                  Math.round(parseFloat(m[3]) * 10)];
  }

  let arms = {};
  if (Object.keys(ends).length) {
    const xs = Object.values(ends).slice().sort((a, b) => a[0] - b[0]);
    const ys = Object.values(ends).slice().sort((a, b) => a[1] - b[1]);
    /* Balagere case-study arm names (from dead_end junction ids) — kept
     * verbatim from sumo_geom.py. */
    arms = { Panathur: xs[0], Varthur: xs[xs.length - 1],
             Sarjapur: ys[0], Kundalahalli: ys[ys.length - 1] };
  }

  const phases = [];
  const stops = {};
  const links = {};

  if (s.includes('<tlLogic')) {
    for (const m of s.matchAll(/<phase duration="([\d.]+)" state="(\w+)"/g)) {
      phases.push([parseFloat(m[1]), m[2]]);
    }
    for (const m of s.matchAll(
      /<connection from="([^"]+)"[^>]*tl="[^"]*"[^>]*linkIndex="(\d+)"/g)) {
      const frm = m[1], idx = parseInt(m[2], 10);
      (links[frm] = links[frm] || []).push(idx);
      const lm = new RegExp('<lane id="' + escapeRe(frm) + '_0"[^>]*shape="([^"]+)"')
        .exec(s);
      if (lm) {
        const last = lm[1].trim().split(/\s+/).pop().split(',');
        stops[frm] = [Math.round(parseFloat(last[0]) * 10),
                      Math.round(parseFloat(last[1]) * 10)];
      }
    }
  }

  return { lanes, arms, phases, stops, links };
}

function escapeRe(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
